libsplinter/src/network/connection_manager/mod.rs

```diff
@@ -0,0 +1,696 @@
+                        if let Err(err) = cm_sender.send(CmMessage::HeartbeatNotifications(notifications)) {
+                            error!("Connection manager has disconnected before shutting down heartbeat monitor {:?}", err);
+                            break;
+                        }
+                }
+            })
+            .unwrap();
+
+        self.join_handle = Some(join_handle);
+    }
+
+    fn shutdown(self) {
+        let join_handle = if let Some(jh) = self.join_handle {
+            jh
+        } else {
+            return;
+        };
+
+        self.running.store(false, Ordering::SeqCst);
+
+        if let Err(err) = join_handle.join() {
+            error!("Failed to shutdown heartbeat monitor gracefully: {:?}", err);
+        }
+    }
+}
+
+#[derive(Clone)]
+pub struct Connector {
+    sender: Sender<CmMessage>,
+}
+
+impl Connector {
+    pub fn connect(&self) -> Result<Strand, ConnectionManagerError> {
+        self.connect_with_timeout(DEFAULT_CONNECTION_TIMEOUT)
+    }
+
+    pub fn connect_with_timeout(&self, timeout: u64) -> Result<Strand, ConnectionManagerError> {
+        let id = Uuid::new_v4().to_string();
+        let (send, recv) = unbounded();
+        match self.sender.try_send(CmMessage::Connect(id.clone(), send)) {
+            Ok(()) => Ok(Strand {
+                id,
+                recv,
+                sender: self.sender.clone(),
+                timeout: Duration::from_secs(timeout),
+            }),
+            Err(_) => Err(ConnectionManagerError::SendMessageError(
+                "The connection manager is no longer running".into(),
+            )),
+        }
+    }
+
+    pub fn subscribe(&self) -> Result<NotificationHandler, ConnectionManagerError> {
+        let id = Uuid::new_v4().to_string();
+        let (send, recv) = unbounded();
+        match self.sender.try_send(CmMessage::Subscribe(id.clone(), send)) {
+            Ok(()) => Ok(NotificationHandler {
+                id,
+                recv,
+                sender: self.sender.clone(),
+            }),
+            Err(_) => Err(ConnectionManagerError::SendMessageError(
+                "The connection manager is no longer running".into(),
+            )),
+        }
+    }
+}
+
+/// Request Handle for the connection manager, used for sending request to
+/// connection manager
+///
+/// ```
+/// use splinter::network::connection_manager::ConnectionManager;
+/// use splinter::mesh::Mesh;
+/// use splinter::transport::inproc::InprocTransport;
+///
+/// let mesh = Mesh::new(1, 1);
+/// let transport = Box::new(InprocTransport::default());
+///
+/// let mut cm = ConnectionManager::new(mesh, transport);
+/// let connector = cm.start();
+///
+/// let strand = connector.connect().unwrap();
+///
+/// let response = strand.request_connection("inproc://foo-bar");
+///
+/// ```
+pub struct Strand {
+    id: String,
+    sender: Sender<CmMessage>,
+    recv: Receiver<Response>,
+    timeout: Duration,
+}
+
+impl Strand {
```

Comment:
I spoke with @peterschwarz and removed the need for `Strand` and moved `request_connection` into `Connector`. I'm not sure if either of those names is appropriate since `Connector` is really just a bidirectional channel.